File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Japan.m
//
//  UIImage+Japan.m
//  CountriesFlags
//

#import "Countries/UIImage+Japan.h"

#import "Extensions/CGSize+Extension.h"

@implementation UIImage (Japan)

+ (UIImage * _Nullable)flagJapanWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:188.0f/255.0f
                                        green:0.0f
                                         blue:45.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));

    [colorRed setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGSizeMidPoint(size)
                                    radius:size.height * 3.0f / 10.0f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Japan.swift
//
//  UIImage+Japan.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Japan "wikipedia"
     
     Generates the flag of Japan.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Japan.
     */
    internal static func flagJapan(with size: CGSize) -> UIImage {
        let colorRed = UIColor(red: 188.0 / 255.0,
                               green: 0.0,
                               blue: 45.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        UIBezierPath(arcCenter: size.midPoint,
                     radius: size.height * 3.0 / 10.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
